File: leetcode/0206.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def reverseList(self, head: Optional[ListNode]) -> Optional[ListNode]:
        if head is None:
            return head
        prev_node, curr_node, next_node = None, head, head.next
        while next_node is not None:
            curr_node.next = prev_node
            prev_node, curr_node, next_node = curr_node, next_node, next_node.next
        curr_node.next = prev_node
        return curr_node

File: leetcode/0206.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode* reverseList(ListNode* head) {
        ListNode *first, *second, *tmp; 
        if(head == NULL)
            return head;
        else
        {
            first = head;
            second = head -> next;
            while(second != NULL)
            {
                tmp = second -> next;
                second -> next = first;
                first = second;
                second = tmp;
            }
        }
        head -> next = NULL;
        return first;
    }
};
